Staging

aggiewb opened the pull request.

aggiewb committed:
Add boilerplate html, css, and js

aggiewb committed:
Add object literals for each shop location with min/max hourly

customers, average cookies per customer, and a function that generates a random number of cookies per hour

aggiewb committed:
Add function that calculates a simulated amount of cookies

per hour for each location
Store each stores' calculatedSimulatedCookiesPerHour in a property of each object literal

aggiewb committed:
Add for loop to set simulatedCookiesPerHour prop with a call

of calculateSimulatedCookiesPerHour()
Correct spelling

aggiewb committed:
Add and call function createStoreLocationHTMLSections(),

which calls the added createH2Element(), createULElement(), and createLIElements
Add name field for each object literal

aggiewb committed:
Add h1 element to html, and add css list-style-type of circle

aggiewb committed:
Add conditional for 12am timeframe in createLIElements(),

add commenting for timeframes for better readability, and remove trailing spaces on js file

aggiewb committed:
Change pm to am in string

BatemanVO commented on js/sales.js:
```diff
@@ -0,0 +1,109 @@
+    averageCookiesPerCustomer: 6.3,
+    randomCustomersPerHour: function() {
+        return Math.ceil(Math.random() * (this.maxHourlyCustomers - this.minHourlyCustomers)) + this.minHourlyCustomers;
+    }
```
We could declare this function outside the context of an object, then assign it to each one, to be a little more DRY. Something like:
![Capture](https://user-images.githubusercontent.com/20998142/84961846-32beeb80-b0ba-11ea-8c94-ce244544b0cd.PNG)

BatemanVO commented on js/sales.js:
```diff
@@ -0,0 +1,109 @@
+
+function calculateSimulatedCookiesPerHour(shopLocation){
+    //using military time
+    var openHours = 20 - 6;
```
Why not just assign the value 14? If it's because you wanted to document that closing hours are 20:00 and opening hours at 06:00, you might want to create some named constants, something like `var OPENING_TIME = 6;` and `var CLOSING_TIME = 20;`, then `var openHours = CLOSING_TIME - OPENING_TIME;`.

Otherwise, they're a bit of what are called [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants).

It does mean caching a handful of values that you don't technically need to, but in this case, the documentation you get from having named variables is worth the trade-off of the negligible efficiency difference.

BatemanVO commented on js/sales.js:
```diff
@@ -0,0 +1,109 @@
+    var newH2Element = document.createElement('h2');
+    newH2Element.setAttribute('class', location.name.replaceAll(" ", "-"));
+    newH2Element.textContent = location.name;
+    document.querySelector('body').append(newH2Element);
```
Perhaps we could `return newH2Element` here, then pass it into `createULElement`? So the `createStoreLocationHTMLSections` could look like:
`var header = createH2Element(storeLocation);`
`createULElement(storeLocation, header);`

That way, on line 80, we don't need to re-query the element we just created.

BatemanVO commented on js/sales.js:
```diff
@@ -0,0 +1,109 @@
+    var h2Element = document.getElementsByClassName(location.name.replaceAll(" ", "-"))[0];
+    var newULElement = document.createElement('ul');
+    newULElement.setAttribute('class', location.name.replaceAll(" ", "-"));
+    h2Element.insertAdjacentElement('afterend', newULElement);
```
Do you end up with the same result if you just do `document.body.appendChild(newULElement);`? I'm unfamiliar with `insertAdjacentElement` (which is a cool discovery!), but at a glance, it seems like we're basically saying, "Create a sibling to the H2 after it", which appending a new child should have the same effect for.

aggiewb replied:
I also applied this change on line 76 which removes a query

BatemanVO commented on js/sales.js:
```diff
@@ -0,0 +1,109 @@
+
+function createULElement(location){
+    var h2Element = document.getElementsByClassName(location.name.replaceAll(" ", "-"))[0];
+    var newULElement = document.createElement('ul');
```
Perhaps we could give these `h2Element`, `ULElement`, etc. variable names to be more descriptive about what they represent - something like `header`, `hoursList`, etc.

BatemanVO commented on js/sales.js:
```diff
@@ -0,0 +1,109 @@
+}
+
+function createLIElements(location){
+    var h2Element = document.getElementsByClassName(location.name.replaceAll(" ", "-"))[1];
```
I know you had an issue with using `querySelector` on an Id that started with a digit, but this way of retrieving an element seems very brittle. Perhaps we can apply a `data-` attribute to each header to query on, or assign them Ids and use `getElementById` instead of `querySelector`, since `getElementById` doesn't use a CSS selector and should be OK retrieving elements based on Ids that start with numbers.

aggiewb replied:
The var h2Element should actually be a ulElement, but I removed this query since I returned the ul element created and passed it to this function.

BatemanVO commented on js/sales.js:
```diff
@@ -0,0 +1,109 @@
+    for(var i = 0; i < location.simulatedCookiesPerHour.length; i++){
+        var newLIElement = document.createElement('li');
+        var hourSoldCookies = location.simulatedCookiesPerHour[i];
+        if(openingHour + i < 12){
```
Since we generate the sum multiple times, it would be a little more efficient to cache `openingHour + i` instead of re-adding it up multiple times.

BatemanVO commented on js/sales.js:
```diff
@@ -0,0 +1,109 @@
+        var hourSoldCookies = location.simulatedCookiesPerHour[i];
+        if(openingHour + i < 12){
+            //1am-11am
+            newLIElement.textContent = (openingHour + i) + "am: " + hourSoldCookies + " cookies";
```
We're being a bit repetitive with how we're assigning the `textContent` property. We could make this code more DRY by perhaps doing something like:
![Capture](https://user-images.githubusercontent.com/20998142/84962973-ea54fd00-b0bc-11ea-8445-14e95c5deef6.PNG)

aggiewb replied:
This won't work since I don't want to display 24am or 13+ pm

BatemanVO commented on sales.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html lang="en">
+    <head>
+        <title></title>
```
We should add a title here (or remove the element altogether, though I think a value would be nice).

BatemanVO commented on sales.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html lang="en">
```
I actually didn't know about the `lang` attribute, but MDN recommends to always include it to help browsers and search engines. Good addition!

BatemanVO commented on js/sales.js:
```diff
@@ -0,0 +1,109 @@
+    var openingHour = 6;
+    for(var i = 0; i < location.simulatedCookiesPerHour.length; i++){
+        var newLIElement = document.createElement('li');
+        var hourSoldCookies = location.simulatedCookiesPerHour[i];
```
Good caching here.

BatemanVO reviewed (commented):
Although the assignments for the ticket are not to dynamically create the HTML, they will be a future requirement, so this is good practice on dynamically generating the page. Well done!

aggiewb committed:
Move cookieLocations to have global access, and declare the

function randomCustomersPerHour() outside each object literal to be called with forEach() on cookieLocations
Update createStoreLocationHTMLSections() to pass in the global cookieLocations
Change openHours to be clearer

aggiewb committed:
Cache each returned newH2Element to be passed to

createULElement()
Update parameters for createULElement() to accept a newH2Element header

aggiewb committed:
Rename variables to be more descriptive, cache each returned

newHoursList to be passed to createHourItems
Remove unused setAttribute of classes

aggiewb committed:
Cache openingHour + i, standardTimeOffSet, and remove

calculation for 12 am/pm and hardcode times into the string

aggiewb committed:
Add html title content

aggiewb committed:
Change appending of newHoursList with a more relatable

method

aggiewb committed:
Remove query and add appendChild() method for each new

header

aggiewb committed:
Add IIFE around all file JS, and add use strict literal expression

aggiewb committed:
Change all double quotes to single quotes, and add line break

for createStoreLocationHTMLSections(cookieLocations) at end of file

aggiewb committed:
Remove unused parameters in createHoursList(), and remove

cached items in createStoreLocationHTMLSections()
